src/dbus-objects/addrconf.c

```diff
@@ -636,80 +614,348 @@ ni_objectmodel_addrconf_forward_request(ni_dbus_addrconf_forwarder_t *forwarder,
+ni_addrconf_drop_forward_new(ni_dbus_addrconf_forwarder_t *forwarder)
+{
+	static const char *info = "forwarding drop lease";
+	ni_addrconf_drop_forward_t *act;
+
+	if (!forwarder || !(act = calloc(1, sizeof(*act))))
+		return NULL;
+
+	act->fwdr = forwarder;
+	act->action.info = info;
+	act->action.exec = ni_addrconf_drop_forward_exec;
+	ni_addrconf_drop_args_init(act->fwdr, &act->args);
+	return act;
+}
+
+static ni_addrconf_drop_collect_t *
+ni_addrconf_drop_collect_cast(ni_addrconf_action_t *action)
+{
+	if (!action || action->exec != ni_addrconf_drop_collect_exec)
+		return NULL;
+	return (ni_addrconf_drop_collect_t *)action;
+}
+
+static inline ni_bool_t
+ni_addrconf_drop_collect_test(ni_addrconf_updater_t *updater, ni_addrconf_drop_collect_t *collect)
+{
+	if (collect->complete)
+		return TRUE;
+
+	if (collect->deadline && !ni_lifetime_left(collect->deadline, &updater->astart, NULL))
+		return TRUE;
+
+	return FALSE;
+}
+
+static int
+ni_addrconf_drop_collect_exec(ni_netdev_t *dev, ni_addrconf_lease_t *lease)
+{
+	ni_addrconf_drop_collect_t *collect;
+
+	if (!dev || !lease || !lease->updater)
+		return -1;
+
+	if (!(collect = ni_addrconf_drop_collect_cast(lease->updater->action)))
+		return -1;
+
+	if (ni_addrconf_drop_collect_test(lease->updater, collect)) {
+		lease->state = NI_ADDRCONF_STATE_RELEASED;
+		return 0;
+	}
+	return 1;
+}
+
+static inline void
+ni_addrconf_drop_collect_free(ni_addrconf_drop_collect_t *act)
+{
+	/* nothing to free inside */
+	free(act);
+}
+
+static ni_addrconf_drop_collect_t *
+ni_addrconf_drop_collect_new(void)
+{
+	static const char *info = "collecting drop event";
+	ni_addrconf_drop_collect_t *act;
+
+	if (!(act = calloc(1, sizeof(*act))))
+		return NULL;
+
+	act->deadline = 10; /* worst case 10 sec no-reply deadline
+			       (buggy/stopped/crashed supplicant) */
+	act->action.info = info;
+	act->action.exec = ni_addrconf_drop_collect_exec;
+	return act;
+}
+
+static ni_bool_t
+ni_addrconf_drop_actions_arm(ni_dbus_addrconf_forwarder_t *forwarder,
+				ni_addrconf_updater_t *updater,
+				const ni_dbus_variant_t *dict)
+{
+	ni_addrconf_drop_forward_t *forward = NULL;
+	ni_addrconf_drop_collect_t *collect = NULL;
+	ni_addrconf_action_t *pos;
+
+	if (!updater)
+		return FALSE;
+
+	if (!(forward = ni_addrconf_drop_forward_new(forwarder)))
+		goto failure;
+
+	ni_addrconf_drop_args_from_variant(forwarder, &forward->args, dict);
+
+	if (!(collect = ni_addrconf_drop_collect_new()))
+		goto failure;
+
+	/*
+	 * Insert the forward action behind address removal
+	 * and append the collect action as the last action.
+	 */
+	pos = ni_addrconf_action_list_find_exec(updater->action,
+				ni_addrconf_action_addrs_remove);
+
+	if (pos && ni_addrconf_action_list_insert(&pos->next, &forward->action)) {
+		if (ni_addrconf_action_list_append(&pos->next, &collect->action))
+			return TRUE;
+	}
+
+failure:
+	ni_addrconf_drop_collect_free(collect);
+	ni_addrconf_drop_forward_free(forward);
+	return FALSE;
```

Comment:
This isn't 100% save, if `if (ni_addrconf_action_list_append(&pos->next, &collect->action))` fails we free `forward` which is already in the list.
On the other hand, there is no way to fail...

Reply:
ni_addrconf_action_list_insert or append cannot fail or in another words, they can fail only when pos would be NULL (what we catch) or when we've failed to allocate forward or collect (both catched before too). We can  do this  instead: 
```diff
@@ -849,9 +850,11 @@ ni_addrconf_drop_actions_arm(ni_dbus_addrconf_forwarder_t *forwarder,
        pos = ni_addrconf_action_list_find_exec(updater->action,
                                ni_addrconf_action_addrs_remove);
 
-       if (pos && ni_addrconf_action_list_insert(&pos->next, &forward->action)) {
-               if (ni_addrconf_action_list_append(&pos->next, &collect->action))
-                       return TRUE;
+       if (pos) {
+               /* insert or append cannot fail -- checked conditions above */
+               ni_addrconf_action_list_insert(&pos->next, &forward->action);
+               ni_addrconf_action_list_append(&pos->next, &collect->action);
+               return TRUE;
        }
 
 failure:
````

Reply:
Should I change to the above variant to avoid confusion about?

Reply:
Adjusted to avoid confusion in https://github.com/openSUSE/wicked/compare/41a0903..0c9aacc